src/vos/vos_io.c

```diff
@@ -2146,37 +2225,102 @@ iod_reserve(struct vos_io_context *ioc, struct bio_iov *biov)
 	D_DEBUG(DB_TRACE, "media %d offset "DF_X64" size %zd\n",
 		biov->bi_addr.ba_type, biov->bi_addr.ba_off,
 		bio_iov2len(biov));
+}
+
+static inline void
+iod_set_sv_addr(struct vos_io_context *ioc, umem_off_t umoff, bio_addr_t *addr)
+{
+	struct vos_sv_addr	*sv_addr;
+
+	D_ASSERTF(ioc->ic_sv_addr_at < ioc->ic_sv_addr_cnt, "sv_at:%u >= sv_cnt:%u\n",
+		  ioc->ic_sv_addr_at, ioc->ic_sv_addr_cnt);
+
+	sv_addr = &ioc->ic_sv_addrs[ioc->ic_sv_addr_at];
+	sv_addr->sa_umoff = umoff;
+	sv_addr->sa_addr = *addr;
+	ioc->ic_sv_addr_at++;
+}
+
+static int
+gang_reserve_sv(struct vos_io_context *ioc, uint16_t media, daos_size_t size,
+		umem_off_t umoff, unsigned int gang_nr)
+{
+	struct vos_irec_df	*irec;
+	struct bio_iov		 biov = { 0 };
+	bio_addr_t		 gaddr = { 0 };
+	daos_size_t		 alloc_sz;
+	uint64_t		 off;
+	char			*gaddr_ptr;
+	int			 i, rc;
+
+	D_ASSERT(gang_nr > 1);
+	D_ASSERT(size > VOS_GANG_SIZE_THRESH);
+
+	irec = (struct vos_irec_df *)umem_off2ptr(vos_ioc2umm(ioc), umoff);
+	gaddr_ptr = vos_irec2data(irec);
+
+	bio_addr_set(&gaddr, DAOS_MEDIA_SCM, umem_ptr2off(vos_ioc2umm(ioc), gaddr_ptr));
+	gaddr.ba_gang_nr = gang_nr;
+	BIO_ADDR_SET_GANG(&gaddr);
+
+	iod_set_sv_addr(ioc, umoff, &gaddr);
+
+	for (i = 0; i < gang_nr; i++) {
```

Comment:
do we need to consider to yield if gang_nr is large? same for the loop in iod_gang_fetch().
MAX gang_nr is 255 now

Reply:
Let's keep it simple for this moment, if it turns out 255 reserve consumes too much CPU, we could add yield then.